Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+
+private theorem mu_property {x y : R} (hn : ∀ n : ℕ, ∃ m < n + 1,
+      μ ((x + y) ^ (n : ℕ)) ^ (1 / (n : ℝ)) ≤ (μ (x ^ m) * μ (y ^ (n - m : ℕ))) ^ (1 / (n : ℝ)))
+    (n : ℕ) : μ ((x + y) ^ (n : ℕ)) ^ (1 / (n : ℝ)) ≤
+      (μ (x ^ mu μ hn n) * μ (y ^ (n - mu μ hn n : ℕ))) ^ (1 / (n : ℝ)) :=
+  (Classical.choose_spec (hn n)).2
+
+
+private theorem mu_le {x y : R} (hn : ∀ n : ℕ, ∃ m < n + 1,
+      μ ((x + y) ^ (n : ℕ)) ^ (1 / (n : ℝ)) ≤ (μ (x ^ m) * μ (y ^ (n - m : ℕ))) ^ (1 / (n : ℝ)))
+    (n : ℕ) : mu μ hn n ≤ n := by
+  simp only [mu, ← Nat.lt_succ_iff]
+  exact (Classical.choose_spec (hn n)).1
+
+private theorem mu_bdd {x y : R} (hn : ∀ n : ℕ, ∃ m < n + 1,
+      μ ((x + y) ^ (n : ℕ)) ^ (1 / (n : ℝ)) ≤ (μ (x ^ m) * μ (y ^ (n - m : ℕ))) ^ (1 / (n : ℝ)))
+    (n : ℕ) : (mu μ hn n : ℝ) / n ∈ Set.Icc (0 : ℝ) 1 := by
+  refine Set.mem_Icc.mpr ⟨div_nonneg (cast_nonneg (mu μ hn n)) (cast_nonneg n), ?_⟩
+  by_cases hn0 : n = 0
+  · rw [hn0, cast_zero, div_zero]; exact zero_le_one
+  · have hn' : 0 < (n : ℝ) := cast_pos.mpr (Nat.pos_of_ne_zero hn0)
+    rw [div_le_one hn', cast_le]
+    exact mu_le _ _ _
+
+private theorem μ_bddBelow (s : ℕ → ℕ) {x : R} (ψ : ℕ → ℕ) :
+    BddBelow {a : ℝ |
+      ∀ᶠ n : ℝ in map (fun n : ℕ => μ x ^ (↑(s (ψ n)) * (1 / (ψ n : ℝ)))) atTop, n ≤ a} := by
+  use 0
+  simp only [mem_lowerBounds, eventually_map, eventually_atTop, ge_iff_le, Set.mem_setOf_eq,
+    forall_exists_index]
+  intro r m hm
+  exact le_trans (rpow_nonneg (apply_nonneg μ _) _) (hm m (le_refl _))
+
+private theorem μ_bddAbove (hμ1 : μ 1 ≤ 1) {s : ℕ → ℕ} (hs : ∀ n : ℕ, s n ≤ n) (x : R)
+    (ψ : ℕ → ℕ) : BddAbove (Set.range fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ))) := by
+  have hψ : ∀ n : ℕ, 0 ≤ 1 / (ψ n : ℝ) := fun _ ↦ by simp only [one_div, inv_nonneg, cast_nonneg]
+  by_cases hx : μ x ≤ 1
+  · use 1
+    simp only [mem_upperBounds, Set.mem_range, forall_exists_index, forall_apply_eq_imp_iff']
+    rintro _ n rfl
+    apply le_trans (rpow_le_rpow (apply_nonneg _ _) (map_pow_le_pow' hμ1 _ _) (hψ n))
+    rw [← rpow_natCast, ← rpow_mul (apply_nonneg _ _), mul_one_div]
+    exact rpow_le_one (apply_nonneg _ _) hx (div_nonneg (cast_nonneg _) (cast_nonneg _))
+  · use μ x
+    simp only [mem_upperBounds, Set.mem_range, forall_exists_index, forall_apply_eq_imp_iff']
+    rintro _ n rfl
+    apply le_trans (rpow_le_rpow (apply_nonneg _ _) (map_pow_le_pow' hμ1 _ _) (hψ n))
+    rw [← rpow_natCast, ← rpow_mul (apply_nonneg _ _), mul_one_div]
+    conv_rhs => rw [← rpow_one (μ x)]
+    rw [rpow_le_rpow_left_iff (not_le.mp hx)]
+    exact div_le_one_of_le₀ (cast_le.mpr (hs (ψ n))) (cast_nonneg _)
+
+private theorem μ_bddAbove' (hμ1 : μ 1 ≤ 1) {s : ℕ → ℕ} (hs : ∀ n : ℕ, s n ≤ n) (x : R)
+    (ψ : ℕ → ℕ) : BddAbove ((fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ))) '' Set.univ) := by
+  rw [Set.image_eq_range]
+  convert μ_bddAbove μ hμ1 hs x ψ
+  ext
+  simp only [one_div, Set.mem_range, Subtype.exists, Set.mem_univ, exists_const]
```

Comment:
```suggestion
  simp [one_div, Set.mem_range, Subtype.exists, Set.mem_univ, exists_const]
```

Reply:
I can change it, but could you explain why you prefer this?

Reply:
Because while non-terminal `simp`s should always be `simp only` to avoid breaking the proof downwards, a terminal `simp` will not be broken by the list of `simp` lemmas being added, and there was a discussion here
https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/request.3A.20discourage.20squeezing.20terminal.20simps.3F

I forgot that this also applies to `simpa`, and I suggested a lot of `simpa only [...] using ...`: can you rever those to `simpa [...] using ...` removing the `only`? Sorry for the fuzz.

Reply:
Done, except for the `simpa only` at line 108 (which breaks if I remove the `only`).